### ISSUE TITLE
Add some way for a system to hide certain grid settings 

### ISSUE THREAD

**Issue by CarlosFdez** [foundryvtt/foundryvtt#10683]
### User Experience

In PF2e, we have explicitly defined square grid diagonals, cone shapes, and grid drawing rules, and we take control of the grid highlight directly. Therefore there may be breakage or unexpected behavior we'd rather not account for if a user starts playing with the settings (there's already some weirdness with cone template shape in v11).

It would be helpful if there were some means to hide the `Square Grid Diagonals`, `Grid-based Template Shapes`, and `Cone Template Shape` core settings. One idea I have is to add a new system manifest setting under grid, but you might have other ideas.

(slightly unrelated, but I noticed that grid-based template shapes was very close to the pf2e shapes, I think the only difference might be the rounding employed, though I have not checked deeply. For reference the shapes are here: https://2e.aonprd.com/Images/Rules/Rules354.png)

**Comment by aaclayton** [comment]
Hey @CarlosFdez thanks for opening this issue. We can consider it. It doesn't sound to me like the best approach though. Instead of fighting against the core grid, it would be better if the PF2E system integrated with the new V12 grid API.

That might entail, for example, defining a system subclass of the `SquareGrid` and overriding some methods in cases where PF2E measurement should work a specific way. This approach would allow you to - if desired - enforce certain diagonal movement rules at the system level and customize the shapes of circular or conal templates.

While I *do not recommend* doing this, if you are intent on suppressing configurability of certain settings you can do that in the system by using the `renderSettingsConfig` hook.

@dev7355608 do you have any additional thoughts or recommendations here?

**Comment by CarlosFdez** [comment]
Thanks for replying. So there are two separate things going on. We do attempt to integrate with the grid api. Currently we subclass MeasuredTemplate#highlightGrid(), to draw the highlight but also to snap the template based on type. There might be newer methods to look at. That said, this ask is more specifically about locking down specific settings to avoid undefined behavior and weird render results (there is a visual bug that exists even in v11 if someone starts fiddling with these settings).

Very likely the approach we'd take otherwise is to do it in renderSettingsConfig the minute we get a bug report from someone misconfiguring it (and remember to reset the setting in init or ready). In fact left to our devices....that's exactly what I'm sure someone would do :P.

If you think its not valuable, that's fine, though I'd definitely keep an eye open. I'm not sure if any system other than D&D5e lets the user select what grid measurement scheme they like. I suspect most systems come to a decision, would configure it a way, and then shame the user if they forgot they touchy.

**Comment by dev7355608** [comment]
I recommend this approach:
```js
// Preventing changes to the setting
Hooks.on("preUpdateSetting", (document, changes) => {
    if ( document.key !== "core.gridDiagonals" ) return;
    if ( Number(changes.value) !== game.system.grid.diagonals ) return false;
});

// Disabling the setting with a tooltip instead of removing it so we don't get users asking in core support why these core settings can't be found anywhere
Hooks.on("renderSettingsConfig", (app, html) => {
    const element = html[0].querySelector(`[name="core.gridDiagonals"]`);
    element.disabled = true;
    element.dataset.tooltip = "Managed and locked by the PF2e system";
});
```

**Comment by CarlosFdez** [comment]
I ended up going with that solution. Thanks for recommending we mark the settings rather than remove outright.

![image](https://github.com/foundryvtt/foundryvtt/assets/1286721/47951680-da21-4f9f-8bf2-75c6e7767c8d)

```ts
Hooks.once("ready", () => {
    // snip
    
    // Enforce certain grid settings (if the settings are the same, onChange isn't triggered)
    game.settings.set("core", "gridDiagonals", 4);
    game.settings.set("core", "gridTemplates", false);
    game.settings.set("core", "coneTemplateType", "round");
    
    // snip
});

Hooks.on("renderSettingsConfig", (_app, $html) => {
    const html = $html[0];
    const lockedSettings = ["core.gridDiagonals", "core.gridTemplates", "core.coneTemplateType"];
    for (const locked of lockedSettings) {
        const element = htmlQuery(html, `div[data-setting-id="${locked}"]`);
        if (!element) continue;

        const controlElement = htmlQuery<HTMLInputElement | HTMLSelectElement>(element, "select, input");
        if (controlElement) {
            controlElement.disabled = true;
            controlElement.dataset.tooltip = "PF2E.SETTINGS.Core.ManagedBySystem";
        }

        const label = htmlQuery(element, "label");
        const lock = createHTMLElement("i", { classes: ["fa-solid", "fa-lock", "fa-fw"] });
        lock.dataset.tooltip = "PF2E.SETTINGS.Core.ManagedBySystem";
        label?.append(" ", lock);
    }
});
```

**Comment by CarlosFdez** [comment]
Some updates after we managed to sit in v12 for a bit, at least for a pf2e specific context.

1) We did not need to lock Square Grid Diagonals. Since we have our own template calculation it only affects rulers. I have a PR to open this setting back up though.
2) We can choose to override grid based template shape's behavior in square scenes only to match pf2e logic, and keep the logic the rest of the same. We can override _computeShape() to keep the nicer looking shapes if we want as well (though we'd need to figure out how to redraw the shape when placed as well, since those placed in the center should expand and those placed in a corner should contract).

**Comment by aaclayton** [comment]
Thanks for the follow-up @CarlosFdez - could you summarize what is your current request or recommendation based on these newest findings?

**Comment by CarlosFdez** [comment]
I don't know. It could be system manifest settings to hide/disable certain options, or some way to alter a setting in the load process (maybe too dangerous because of modules). It might be fine to encourage systems to handle it like we did, as grid settings are likely just the tip of the iceberg of settings systems may want to modiify.

As of now, it looks like in pf2e we're opening up grid diagonals, overriding grid templates to a new meaning (pf2e vs grid based or pf2e vs euclidian, haven't decided), and locking down flat vs round cones (until we can situationally enable settings based on other settings).

I think it's fine to close the issue. Thanks again for your time, as well as dev for the recommended solution.
